File: bindings/flutter/open_eyes/ios/Classes/OpenEyesPlugin.swift
import Flutter
import UIKit

/// Flutter plugin for open-eyes — bridges MethodChannel/EventChannel to the Rust engine.
///
/// The hot path (camera frames) NEVER goes through Flutter.
/// Native AVCaptureSession → Rust directly.
/// Flutter only sees geometric results via EventChannel.
public class OpenEyesPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var engine: OpenEyesEngine?
    private var cameraSource: PhoneCameraSource?
    private var eventSink: FlutterEventSink?

    // Timer for polling Rust events and forwarding to Flutter
    private var pollTimer: Timer?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "com.cambriantech.open_eyes/engine",
            binaryMessenger: registrar.messenger()
        )
        let eventChannel = FlutterEventChannel(
            name: "com.cambriantech.open_eyes/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = OpenEyesPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            engine = OpenEyesEngine()
            startEventPolling()
            result(nil)

        case "startCamera":
            guard let engine = engine else { result(FlutterError(code: "NO_ENGINE", message: "Call initialize first", details: nil)); return }
            let args = call.arguments as? [String: Any] ?? [:]
            let cameraId = args["cameraId"] as? Int ?? 0
            cameraSource = PhoneCameraSource(cameraId: UInt32(cameraId), engine: engine)
            cameraSource?.start()
            result(nil)

        case "stopCamera":
            cameraSource?.stop()
            cameraSource = nil
            result(nil)

        case "getMotion":
            result(Double(engine?.motionMagnitude ?? 0.0))

        case "getFrameCount":
            result(Int(engine?.frameCount ?? 0))

        case "setIntrinsics":
            guard let engine = engine, let args = call.arguments as? [String: Any] else { result(nil); return }
            engine.setIntrinsics(
                cameraId: UInt32(args["cameraId"] as? Int ?? 0),
                fx: args["fx"] as? Double ?? 0,
                fy: args["fy"] as? Double ?? 0,
                cx: args["cx"] as? Double ?? 0,
                cy: args["cy"] as? Double ?? 0,
                width: UInt32(args["width"] as? Int ?? 0),
                height: UInt32(args["height"] as? Int ?? 0)
            )
            result(nil)

        case "enableAR":
            // TODO: start ARSession, feed ARFrame.capturedImage + camera.transform to engine
            result(false)

        case "dispose":
            stopEventPolling()
            cameraSource?.stop()
            cameraSource = nil
            engine = nil
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - Event streaming

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func startEventPolling() {
        // Poll Rust events at 30Hz and forward to Flutter as geometry
        pollTimer = Timer.scheduledTimer(withTimeInterval: 1.0 / 30.0, repeats: true) { [weak self] _ in
            self?.pollEvents()
        }
    }

    private func stopEventPolling() {
        pollTimer?.invalidate()
        pollTimer = nil
    }

    private func pollEvents() {
        guard let engine = engine, let sink = eventSink else { return }

        // Also send periodic motion update even without explicit events
        let motion = engine.motionMagnitude
        if motion > 0.01 {
            sink([
                "type": 0, // motion
                "cameraId": 0,
                "value": Double(motion),
                "position": [0.0, 0.0, 0.0],
                "timestamp": Date().timeIntervalSince1970,
            ] as [String: Any])
        }
    }
}

File: bindings/flutter/open_eyes/android/src/main/kotlin/com/cambriantech/open_eyes/OpenEyesPlugin.kt
package com.cambriantech.open_eyes

import android.os.Handler
import android.os.Looper
import com.cambriantech.openeyes.OpenEyesEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Flutter plugin for open-eyes — bridges MethodChannel/EventChannel to the Rust engine.
 *
 * The hot path (camera frames) NEVER goes through Flutter.
 * Native CameraX → Rust directly via JNI.
 * Flutter only sees geometric results via EventChannel.
 *
 * Works on: Android phones, tablets, Meta Quest headsets.
 */
class OpenEyesPlugin : FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private var channel: MethodChannel? = null
    private var eventChannel: EventChannel? = null
    private var eventSink: EventChannel.EventSink? = null
    private var engine: OpenEyesEngine? = null
    private val handler = Handler(Looper.getMainLooper())
    private var pollRunnable: Runnable? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, "com.cambriantech.open_eyes/engine").also {
            it.setMethodCallHandler(this)
        }
        eventChannel = EventChannel(binding.binaryMessenger, "com.cambriantech.open_eyes/events").also {
            it.setStreamHandler(this)
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel?.setMethodCallHandler(null)
        eventChannel?.setStreamHandler(null)
        stopEventPolling()
        engine?.close()
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> {
                engine = OpenEyesEngine()
                startEventPolling()
                result.success(null)
            }
            "startCamera" -> {
                // TODO: create CameraX ImageAnalysis, bind to engine
                // val cameraId = call.argument<Int>("cameraId") ?: 0
                result.success(null)
            }
            "stopCamera" -> {
                result.success(null)
            }
            "getMotion" -> {
                result.success((engine?.motionMagnitude ?: 0f).toDouble())
            }
            "getFrameCount" -> {
                result.success(engine?.frameCount?.toInt() ?: 0)
            }
            "setIntrinsics" -> {
                engine?.setIntrinsics(
                    cameraId = call.argument<Int>("cameraId") ?: 0,
                    fx = call.argument<Double>("fx") ?: 0.0,
                    fy = call.argument<Double>("fy") ?: 0.0,
                    cx = call.argument<Double>("cx") ?: 0.0,
                    cy = call.argument<Double>("cy") ?: 0.0,
                    width = call.argument<Int>("width") ?: 0,
                    height = call.argument<Int>("height") ?: 0,
                )
                result.success(null)
            }
            "enableAR" -> {
                // TODO: start ARCore session
                result.success(false)
            }
            "dispose" -> {
                stopEventPolling()
                engine?.close()
                engine = null
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // Event streaming to Flutter
    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    private fun startEventPolling() {
        pollRunnable = object : Runnable {
            override fun run() {
                pollEvents()
                handler.postDelayed(this, 33) // ~30Hz
            }
        }
        handler.post(pollRunnable!!)
    }

    private fun stopEventPolling() {
        pollRunnable?.let { handler.removeCallbacks(it) }
        pollRunnable = null
    }

    private fun pollEvents() {
        val sink = eventSink ?: return
        val motion = engine?.motionMagnitude ?: return

        if (motion > 0.01f) {
            sink.success(mapOf(
                "type" to 0, // motion
                "cameraId" to 0,
                "value" to motion.toDouble(),
                "position" to listOf(0.0, 0.0, 0.0),
                "timestamp" to (System.currentTimeMillis() / 1000.0),
            ))
        }
    }
}
